hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/ToOneAttributeMapping.java

```diff
@@ -1637,7 +1637,7 @@ else if ( isNullable || hasNotFoundAction() ) {
 						return false;
 					}
 
-					if ( navigablePath.equals( np.getParent() ) ) {
+					if ( navigablePath.getIdentifierForTableGroup().equals( np.getParent().getIdentifierForTableGroup() ) ) {
```

Comment:
`navigablePath` has an alias while `np.getParent()` has not so ` navigablePath.equals( np.getParent())` is false while `navigablePath.getIdentifierForTableGroup().equals( np.getParent().getIdentifierForTableGroup() )` is true

Reply:
I think I'd prefer a dedicated method for this instead i.e.

```java
public boolean isSameIgnoreAlias(NavigablePath p) {
    return Objects.equals( parent, p.parent ) && localName.equals( p.localName );
}
```

but I'm not sure how to name this best.

Reply:
Does the same problem potentially exist wrt the parent?  Meaning, can the parent's also be aliased and need to be checked in similar manner?

Also, I'd flip that check.  Locally checking one String against another (localName) is going to be better performing compared to recursively walking parents first.  Not to mention the question about alias on parent.

Reply:
I didn't debug this yet, but judging from the test case, I think this runs into an optimization where an explicitly joined table can be reused for an implicit join (to avoid doing 2 joins in SQL), so AFAIU this could happen for parents also. In that case, I think I'd go with this:

```java
public boolean isSameIgnoreAlias(NavigablePath p) {
    return p != null && localName.equals( p.localName ) && parent == null ? p.parent == null : parent.isSameIgnoreAlias( p.parent );
}
```

Reply:
The name is not terrible, but "ignore" is odd English.  How about `#isSameIgnoringAlias`?  Or `#isSameSansAlias`?

Really the intent seems to be that they are the same path, so maybe `#pathsMatch`?  

But this makes me wonder... won't this potentially match incorrectly?  E.g. if I have `Person(a).address.city` and `Person(b).address.city`, these will be the "same" according to these checks iiuc.  Or are these checks already "scoped" to the proper sub-paths?

Reply:
I'm pretty sure this is properly "scoped" already, but isn't the `localName` of the root path the alias anyway? If it is not, I would prefer suggest we make sure this method compares aliases of the root `NavigablePath`. The name, `pathsMatch` sound good to me.

Reply:
I think `navigablePath.getLocalName().equals( np.getParent().getLocalName()` can work.

Reply:
> I'm pretty sure this is properly "scoped" already, but isn't the `localName` of the root path the alias anyway? 

No.  E.g., for `from Person p ...` the root path is `Person(p)`.

> If it is not, I would prefer suggest we make sure this method compares aliases of the root `NavigablePath`. 

Its not just root though.  E.g. `from Person p join p.address a1 join p.address a2`

I'm not understanding enough the conditions where we want to ignore the aliases though to make the most informed decisions about how to accomplish that.  Just pointing out that ignoring aliases will cause problems when we ignore them.

Reply:
paired with @beikov and updated the solution.